### PR TITLE
Grid component react

### PR DESCRIPTION
**Pull request by zachhardesty7**
This is my first major PR (for Carbon & in general) so any feedback is appreciated! My code focuses heavily on readability and simplicity while mimicking design patterns from other React components, Carbon and otherwise.

Add layout helpers for managing a Grid alongside the Carbon Design System layout guidance. Namely, this PR adds support for Grid, Row, and Col components. Grid and Row are roughly the same layout containers, supporting various modes for "condensed" or "no gutter" moments. Column provides support for specifying responsive breakpoints using individual props, as demonstrated by the storybook story and tests. (mostly copied description)

All of this work started before @joshblack's PR #3892. I should have opened an issue sooner... I've mostly incorporated all of his PR, but notable differences of this PR include:

- `noGutterLeft` & `noGutterRight` merged into `noGutter` as "left" and "right"
- no object syntax, all props separate. this decision was made to mimic some very popular libraries ([Material UI](https://material-ui.com/api/grid/), [Ant Design](https://ant.design/components/grid/#Col), [React Bootstrap](https://react-bootstrap.netlify.com/layout/grid/#col-props)). I do also like the object syntax so that could be merged.
- no array syntax ^, but makes helper functions simpler & cleaner
- calls `classnames` function with params & avoids object key calcs for simplicity
- a lot of focus on code readability
- `propTypes` defines supported column span as specific numbers
- Grid supports `noGutter`
- all components don't define any variables in function body
- includes all of the same tests plus a few extra and they all use shallow rendering
- includes dynamic knob-based stories while also preserving the easy to copy and paste ones. maybe too many stories?

#### Changelog

**New**

- add Grid component with supporting Grid.Row (and GridRow) and Grid.Col (and GridCol)

**Changed**

- minor Storybook configs to support [new CSF](https://medium.com/storybookjs/component-story-format-66f4c32366df) 

**Removed**

#### Testing / Reviewing

The styles and examples come roughly from the [Grid elements preview](https://carbon-elements.netlify.com/grid/examples/preview/).

### CODE REVIEW

**Comment by joshblack** [comment]
Hi @zachhardesty7! 👋 Thanks so much for taking the time to put this together! I will definitely take a look at this tomorrow!! 🥳

**Comment by joshblack** [comment]
Taking a quick look, any chance we could remove support for `condensed` and `noGutter`? There might be changes in terminology for these so they might not be settled enough to go into stable.

Definitely understand the prop changes in part, I think `span` and `offset` were patterns brought over from styled system-esque APIs in my earlier PR. With this named approach, I think something that can feel awkward are the `offset*` props that come up, as a result (`offsetSm`, `offsetMd`, etc). It seems like React bootstrap handles these the best but even within their model one can't do a shorthand notation for offsets which is a bummer.

Are there any other examples that you'd recommend looking at for an ideal grid API here? Seems like we would want:

- Auto column support
- Span `n` columns per breakpoint
- Offset `n` columns per breakpoint

**Comment by zachhardesty7** [comment]
@joshblack 

I only implemented those since the Grid component/element package demo and the outputted CSS support them. Does [the doc here](https://github.com/carbon-design-system/carbon/blob/master/packages/grid/docs/sass.md#carbon--no-gutter-mixin) (the red X) indicate that the classes output by the SCSS shouldn't be used? It wasn't quite clear to me whether they're internal or not. It's an easy change though!

I'm not a huge fan of the offset stuff as separate props either, feels a bit messy. I suppose it could be done via a single `offset` prop in object notation: `offset={ sm: 1, lg: 10, max: 12 }`. My concern there was not having symmetry with the span props (sm, md). And then should there be two separate ways for the span props via object notation (eg `span={ sm: 1, lg: 10, max: 12 }` to maintain symmetry? 

This implementation handles auto via the string `"auto"`. Bootstrap supports that and a value of `true`.

In personal and freelance projects, it feels like I've rarely used props as a part of the Grid, but maybe it would benefit us to have some numbers behind how often props are used? I haven't done anything with parsing Github code usage, but I suspect the most popular APIs should mirror usage. Let me look into it a bit further and respond with some more detailed resources of what I find!

**Comment by zachhardesty7** [comment]
@joshblack 
Okay, so `{breakpoint}Offset` style props are not good. The better question is what way "offset" should be handled. Here are what some example React libraries do:

- [Bootstrap React](https://react-bootstrap.netlify.com/layout/grid/#col-props)/[Reactstrap](https://reactstrap.github.io/components/layout/) - no offset prop, must use obj syntax for each breakpoint prop (`sm`, `md`)
- [Antd](https://ant.design/components/grid/#Col) - offset prop applies to all screen sizes, obj syntax for breakpoint specific offset
- [MD](https://material-ui.com/api/grid/#props) - no offset prop
- [SemanticUI](https://react.semantic-ui.com/collections/grid/) - no offset prop

Antd has a cool way of doing it, but since they base it on the Bootstrap style anyway, it makes the most sense to get rid of the offset prop entirely and just support `sm={4}`, `sm={true || false || "auto"}` and also `lg={{ span: 6, offset: 2 }}`. Shouldn't take long to incorporate that here. 

---

### Disqualified examples / research data conclusions
(note: repos without a significant commit within about the last year are excluded)
#### Misc
[React MD - too many props](https://react-md.mlaursen.com/components/grids?tab=1)
[Rebass - CSS-in-JS](https://rebassjs.org/)
[Foundation - can't find prop docs?](https://bit.dev/digiaonline/react-foundation/xy-grid)
[Pinterest Gestalt - too many props](https://pinterest.github.io/gestalt/#/Column) 

#### Seem much more like Carbon DataTable
[Grid Layout](https://github.com/strml/react-grid-layout#responsive-usage), [AG Grid](https://www.ag-grid.com/react-grid/), [Griddle](http://griddlegriddle.github.io/Griddle/)

#### No flexbox-style responsive grid component (e.g. uses CSS Grid style or no Grid included)
[Blueprint](https://blueprintjs.com/docs/#core/variables.grid-system), [React Toolbox](http://react-toolbox.io), [Belle](http://nikgraf.github.io/belle), [Khan Academy](http://khan.github.io/react-components/), [Hedron](https://github.com/garetmckinley/hedron), [Atlas Kit](https://atlaskit.atlassian.com/docs/getting-started), [MS Office Fabric](https://developer.microsoft.com/en-us/fabric#/controls/web), [Grommet](https://v2.grommet.io/grid)

#### Grid component is not responsive / does not support media queries
[Onsen UI](https://onsen.io/v2/api/react/Col.html), [React Virtualized](https://github.com/bvaughn/react-virtualized/blob/master/docs/Grid.md)

**Comment by joshblack** [comment]
Hey @zachhardesty7! Sorry for the delay, just got back from vacation 🌴 haha. Would you want to hop on a webex to chat about the details above? Would love to learn more 😄

**Comment by joshblack** [comment]
@zachhardesty7 definitely a fan of the Antd approach after taking a look, totally get what you were getting at above. I think the breakpoint props that you have already that can be a number or object would work out perfect 💯 I think if we can get a PR with those props and the corresponding components it'd be a great start before tackling things for gutter/hanging type 👍

File: packages/react/package.json
Comment (joshblack) [inline packages/react/package.json]: For storybook updates should we pull these into a separate PR?

File: packages/react/src/components/Grid/Grid.js
Comment (joshblack) [inline packages/react/src/components/Grid/Grid.js]: I think we'll need to make sure we add this to `react/package.json` 👍

File: packages/react/src/components/Grid/Grid.js
Comment (joshblack) [inline packages/react/src/components/Grid/Grid.js]: Quick style nits:

```jsx
export function Grid({
  as: BaseComponent = 'div',
  condensed = false
  fullWidth = false,
  className: containerClassName,
  ...rest,
}) {
  const className = cx(containerClassName, {
    [`${prefix}--grid`]: true,
    [`${prefix}--grid--condensed`]: condensed,
    [`${prefix}--grid--full-width`]: fullWidth,
  });
  return <BaseComponent className={className} {...rest} />;
}
```

File: packages/react/src/components/Grid/Grid.js
Comment (joshblack) [inline packages/react/src/components/Grid/Grid.js]: Maybe `elementType` might be helpful here? Always forget which one is for elements lol

```suggestion
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.elementType]),
```

File: packages/react/src/components/Grid/Grid.js
Comment (joshblack) [inline packages/react/src/components/Grid/Grid.js]: Style-nit: comments need to be formatted like:

```
/**
 * ...
 */
```

In the past it was because of storybook docgen generation but might not be the case anymore 🤔

File: packages/react/src/components/Grid/Grid.js
Comment (joshblack) [inline packages/react/src/components/Grid/Grid.js]:
```suggestion
```

We should probably save `noGutter` for future PRs!

File: packages/react/src/components/Grid/Grid.js
Comment (joshblack) [inline packages/react/src/components/Grid/Grid.js]:
```suggestion
```

File: packages/react/src/components/Grid/Grid.js
Comment (joshblack) [inline packages/react/src/components/Grid/Grid.js]:
```suggestion
export const Row = ({
```

Also should follow same style notes from above 👍

File: packages/react/src/components/Toolbar/Toolbar-story.js
Comment (joshblack) [inline packages/react/src/components/Toolbar/Toolbar-story.js:32]: If we could leave `[Deprecated]` first that'd be great 👍

File: packages/react/src/components/Grid/index.js
Comment (joshblack) [inline packages/react/src/components/Grid/index.js]:
```suggestion
```

I bet we could just use `export *` and drop the `import` and `export default`!

File: packages/grid/scss/_mixins.scss
Comment (joshblack) [inline packages/grid/scss/_mixins.scss]: Could we move the 0 span enhancements under a separate PR? 🙏 Definitely a great feature but might help to pull it out 👍

File: packages/react/src/components/Grid/Grid.js
Comment (joshblack) [inline packages/react/src/components/Grid/Grid.js]: Since some folks might be using these components in different grid modes (12 or 16) would it make sense to loosen this prop type to `PropTypes.number`?

File: packages/react/src/tools/array.js
Comment (joshblack) [inline packages/react/src/tools/array.js]: I bet we could just inline this in the module instead of pulling it out 👍

**Review by joshblack — CHANGES_REQUESTED** [review]

File: packages/react/package.json
Comment (zachhardesty7) [reply packages/react/package.json]: The only reason that I included the updates was due to a single bug that was fixed in 5.2.2, PR here: https://github.com/storybookjs/storybook/pull/8006.
Essentially, I wanted a "blank" default option in the knobs to pass undefined but I had to use the string "undefined" and parse it due to the above.

Scratch that. I never removed the workaround so it could be split out easily.

File: packages/react/src/components/Grid/index.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/index.js]: This appears to disallow a default import for the user (`import Grid from './Grid';`)

File: packages/react/package.json
Comment (zachhardesty7) [reply packages/react/package.json]: resolved via 573727e

File: packages/react/src/components/Grid/Grid.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/Grid.js]: oops, missed this one originally. resolved in 35d8769.

File: packages/react/src/components/Grid/Grid.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/Grid.js]: honestly, looking back at it, this syntax is easier to understand than what I wrote, not sure why I felt the need to change it... 🤷‍♂
279bc1d

File: packages/react/src/components/Grid/Grid.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/Grid.js]: seems to work both ways now! I feel like there should be a way to enforce this but maybe you'd have issues with inline type overrides.
regardless, fixed here 279bc1d

File: packages/react/src/components/Grid/Grid.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/Grid.js]: 021e1a0

File: packages/react/src/components/Grid/Grid.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/Grid.js]: 279bc1d

File: packages/react/src/components/Grid/Grid.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/Grid.js]: 279bc1d

File: packages/react/src/tools/array.js
Comment (zachhardesty7) [reply packages/react/src/tools/array.js]: 179e50b

File: packages/react/src/components/Toolbar/Toolbar-story.js
Comment (zachhardesty7) [reply packages/react/src/components/Toolbar/Toolbar-story.js:32]: The issue here is that Storybook loads all of the stories sequentially, resulting in an alphabetic ordering based on file name. This puts the `Grid	` component at the end of the list when it's loading stories in the new and old formats and thus forcefully sorts it. This incorrectly puts toolbar with the other "D" components. https://github.com/carbon-design-system/carbon/pull/4894/files/acd0d48408f4ad561cf705e69f04b631543118dc#diff-2e6866a078d3aa8cca27fed58050c6cdR35-R36
Referenece: 
https://storybook.js.org/docs/configurations/options-parameter/#sorting-stories

File: packages/react/src/components/Grid/index.js
Comment (joshblack) [reply packages/react/src/components/Grid/index.js]: @zachhardesty7 I think that's totally fine internal to the project, the public way to consume this should be:

```js
import { Grid } from 'carbon-components-react';
```

Internally, we can just write it as:

```js
import { Grid, Row, Cell } from '../components/Grid';
```

File: packages/react/src/components/Grid/Grid.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/Grid.js]: d591165

File: packages/react/src/components/Grid/index.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/index.js]: Whoops, of course the user would never import from `'./Grid'`... d12c3ef

File: packages/grid/scss/_mixins.scss
Comment (zachhardesty7) [reply packages/grid/scss/_mixins.scss]: reverted in 9afd776, d6e6bf9, 54d56d5.

PR here #5052

(there's probably a more eloquent way to transfer 2 commits to a new PR that I'm unfamiliar with, but just creating new commits and referencing this is wayyyy quicker 😄)

File: packages/react/src/components/Grid/Grid.js
Comment (zachhardesty7) [reply packages/react/src/components/Grid/Grid.js]: If I'm not mistaken, since `VALID_COL_WIDTHS` is generated from the `@carbon/layout` package, wouldn't they just be the right value of up to 12 or 16 depending on the user's settings? Maybe that whole check is getting too fancy and it should all be replaced with `number`?

**Comment by joshblack** [comment]
Hey @zachhardesty7! 👋 Thanks so much for taking the time to make updates 🙏 One quick note, would you mind checking the "allow edits from maintainers" checkbox that should be to the right of the PR (under the labels sections). This will help us update this from master and also clean up anything before merging in 😄

**Comment by zachhardesty7** [comment]
@joshblack It appears that the box was already checked, but regardless, it's checked currently! 👍

**Comment by joshblack** [comment]
Thanks @zachhardesty7!

**Comment by joshblack** [comment]
bump @vpicone if you wanted to take a look in case you all wanted to use it in theme/website 👀. I know you all have a component already but would be great if we consolidated.

**Comment by joshblack** [comment]
@zachhardesty7 would you mind if I push some updates to your PR here or would you prefer if I opened up a PR to merge into this branch? I just wanted to double-check with you to see how you felt about the changes, didn't want it to feel like we were disregarding any of your work here as it's all awesome! Thanks again for taking the time to contribute!

**Comment by zachhardesty7** [comment]
@joshblack Whatever makes the most sense / seems the most reasonable to you is fine by me!

**Comment by joshblack** [comment]
@zachhardesty7 just pushed some updates, would love to know what you think! Tried to do the following:

- Split out components into separate files (`Grid`, `Row`, and `Column`)
- Moved tests that you originally had over into separate files under `__tests__` and updated to use our new test helper
  - Tried to use existing tests alongside adding some other checks or doing matching not on exact selector
- Updated story to include the templates since they seemed really solid, don't think anything changed there
- Updated the `Col` component and how it derived the relevant class names, let me know if I missed any functionality here!

**Comment by joshblack** [comment]
FYI @alisonjoseph too for Grid components 👀  Relevant storybook story for usage: https://github.com/carbon-design-system/carbon/pull/4894/files#diff-d9adc48479ce140b1bc043298685114c

File: packages/react/src/components/Grid/Grid-story.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Grid-story.js]: I think this and the below story are both really cool and fun to play around with, but I trust your knowledge that they're probably superfluous! Perhaps it's better omitted like this and left as a learning exercise of Storybook Knobs for myself.

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js]: The object format should probably support booleans too.
```
span: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
```

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js]: Wouldn't we only want to support the value `true`? I feel like `false` shouldn't create an auto-sized column

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js:42]: The object format should probably support booleans too
```suggestion
     span: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
```

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js]: Wouldn't we only want to support the value `true`? I feel like `false` shouldn't create an auto-sized column

File: packages/react/src/components/Grid/__tests__/Column-test.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/__tests__/Column-test.js]: same as above about not supporting `false`
```suggestion
  it('should set the column span to auto if `true` is provided for a breakpoint', () => {
```

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js]: types should include boolean: `{boolean|number}`

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js]: should include boolean:
```suggestion
 * @param {Array<boolean|number|Breakpoint>} breakpoints
```

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js:142]: `span` should support a boolean within the obj format, using the format in the later commit or just:
```
if (span === true) {
  classNames.push(`${prefix}--col-${name}`);
  continue;
}
```

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js:28]: shouldn't the `col` class always be applied like in `Grid` and `Row`?
```suggestion
    [`${prefix}--col`]: true,
```

File: packages/react/src/components/Grid/Column.js
Comment (zachhardesty7) [inline packages/react/src/components/Grid/Column.js:131]: last comment: there appears to be a minor difference between `${prefix}--col-${name}` and `${prefix}--col-${name}--auto` of respectively having `width: 100%` and `width: auto`. I presume we should be using the class with auto but if this change was intentional, I'm curious about why!

(note: this *also* applies to the usage of a boolean within the prop in obj format)

File: packages/react/src/components/Grid/Column.js
Comment (joshblack) [reply packages/react/src/components/Grid/Column.js:28]: @ZacharyStair I believe all of the grid column classes are setup this way and `bx--col` is a short-hand for the auto-span!

File: packages/react/src/components/Grid/Column.js
Comment (joshblack) [reply packages/react/src/components/Grid/Column.js]: @zachhardesty7 it should be caught by the earlier `if (!breakpoint)`!

File: packages/react/src/components/Grid/Column.js
Comment (joshblack) [reply packages/react/src/components/Grid/Column.js:131]: Oh wild, let me take a look real quick to see why 👀 Thanks for the catch!

Comment (joshblack) [reply packages/react/src/components/Grid/Column.js:131]: So far couldn't really find anything 🤔 Will keep digging but will move on with the `width: 100%` unless there is a good reason to use the auto classes.

**Comment by joshblack** [comment]
Updated @zachhardesty7! Thanks again!

**Review by abbeyhrt — APPROVED** [review]
This looks good to me! Thank you @zachhardesty7 for taking the time to do this! 🎉

**Review by aledavila — APPROVED** [review]

**Pull request merged**
